src/executables/sysrepo-plugind.c

```diff
@@ -527,6 +534,34 @@ main(int argc, char **argv)
         goto cleanup;
     }
 
+    /* open and lock pid file before plugins are loaded, but don't update it yet 
+     * plugins may fail, and do some initialization that must finish before we are
+     * considered up and running. No need to create a pid file if we're in debug
+     * mode (we're not a daemon)
+     */
+
+    if (debug && pidfile) {
+        SRPLG_LOG_WRN("sysrepo-plugind", "Ignoring pidfile in debug mode.");
+    }
+
+    if (!debug && pidfile) {
+        /* make sure we are the only instance - lock the PID file and write the PID */
+        pidfd = open(pidfile, O_RDWR | O_CREAT, 0640);
```

Comment:
You can use `O_TRUNC` here and save a later `ftruncate()`. Also, since the comments suggest that the code is trying to only write out a PID once all plugins have loaded, that's one more reason to truncate  that file at the earliest possible time. (But the code is not syncing the IO anyway, so it probably doesn't matter that much.)

Reply:
I don't want to trigger observers like `finit` (an init package: https://github.com/troglobit/finit) that allows the use pid files for conditions.

Reply:
Correction: I do want to trigger observers like that, but not until _after_ plugins have been initialized.